src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs

```diff
@@ -693,6 +700,85 @@ protected override void SetItem(int index, string item)
             {
                 base.SetItem(index, string.IsNullOrEmpty(item) || item == "*.*" ? "*" : item);
             }
+
+            internal string[] GetFilters() => ((ImmutableStringList)Items).Items;
+
+            /// <summary>
+            /// List that maintains its underlying data in an immutable array, such that the list
+            /// will never modify an array returned from its Items property. This is to allow
+            /// the array to be enumerated safely while another thread might be concurrently mutating
+            /// the collection.
+            /// </summary>
+            private sealed class ImmutableStringList : IList<string>
+            {
+                public string[] Items = Array.Empty<string>();
+
+                public string this[int index]
+                {
+                    get
+                    {
+                        string[] items = Items;
+                        if ((uint)index >= (uint)items.Length)
+                        {
+                            throw new ArgumentOutOfRangeException(nameof(index));
+                        }
+                        return items[index];
+                    }
+                    set
+                    {
+                        string[] clone = (string[])Items.Clone();
+                        clone[index] = value;
```

Comment:
What does this do when the list is still `Array.Empty` ?

Reply:
Clone always returns a clone; it doesn't recognize Array.Empty as being special. But even if it did, it wouldn't matter here.  There's no possible value of `index` that would result in the array being mutated. The setter is also guarded by a length check in `Collection<T>`, so we wouldn't get here with an invalid index anyway:
https://github.com/dotnet/corefx/blob/1b3a095c277476b746e6fb9884662cf12334c6a1/src/Common/src/CoreLib/System/Collections/ObjectModel/Collection.cs#L54-L57